calenderのプログラム(Ruby)の実装（複雑な仕様に対応, 一時的にアップデート保留中）

yysd94 opened the pull request:
## 実施課題

[[プラクティス カレンダーのプログラム\(Ruby\) \| FBC](https://bootcamp.fjord.jp/practices/194)](https://bootcamp.fjord.jp/practices/194)

## 実施タスク

02.calendar/cal.rbを追加しました。

## レビューしていただきたい項目

下記の仕様通りにコンソール上に表示が行われること。

### 必須要件リスト

- ruby-practice/02.calendarのディレクトリに移動し、`./cal.rb`で実行できる
- mで月を、-yで年を指定できる　（少なくとも1970年から2100年までは正しく表示される）
- yのみ指定した場合、"一年分のカレンダーを表示する機能は未実装です"と表示される。
- 引数を指定しない場合は、今月・今年のカレンダーが表示される
- MacやWSLに入っているcalコマンドと同じ見た目になっている

※　補足事項
自分の実行環境のcalコマンドでは、-yを指定すると-mオプションを指定していても１年分のカレンダーが表示される実装となっていました。今回実装したプログラムでは、課題の必須要件を優先して、特定の月だけを表示するようにしています。


### 実装した歓迎要件リスト

- 今日の日付の部分の色が反転する(背景色と文字色が入れ替わる)
- どのような引数が与えられようが、cal コマンドと同じ表示結果になる
必須要件の仕様が崩れない範囲で、下記についてcalコマンドの仕様に対応させました。
    - コマンドラインの入力パターンに下記の２パターンでの入力を対応
    cal.rb [-y] [[month] year]
    cal.rb [-m month] [year]
    - 月の指定に英単語での入力を許容する機能。
    - 年を指定しないで、mオプションで月のみ指定する場合に、数字の末尾にf(またはp)をつけることで指定された月番号でかつ現在基準で直後にやってくるもの(または直前のもの)を表示できる機能
    - invalidな値を入れるとエラーメッセージが表示される機能

比較したcalコマンドのバージョンは下記の通りです。

```
❯ apt list --installed | grep ncal
ncal/stable,now 12.1.8 amd64 [installed]

```

## セルフチェックで用いた実行環境

```
❯ ruby --version
ruby 3.3.1 (2024-04-23 revision c56cd86388) [x86_64-linux]

❯ apt list --installed | grep ncal
ncal/stable,now 12.1.8 amd64 [installed]

```

## 備考

rubocopは不要と記載ありましたので、通していません。

以上、宜しくお願いいたします！

yysd94 committed:
西暦、月を引数に指定してその月のカレンダーをコンソールに表示する関数を実装しました。

西暦、月を引数に指定しない場合のデフォルト値はプログラムの実行日の日付としています。
引数が有効な数値であるかどうかのバリデーションは未実装です。

yysd94 committed:
月のカレンダーを表示する関数の名前を、その関数の役割が分かるように修正しました。

yysd94 committed:
コマンドラインオプションの機能に必要な処理について説明するコメントを配置しました。

yysd94 committed:
コマンドライン上に入力されたオプションと引数から指定された年・月を取得する機能を実装しました。

yysd94 committed:
-m、-y オプション、コマンドライン引数を入力して、特定の年月のカレンダーを表示する機能を実装しました。

今回実装したものは、下記の使い方に対応しています。
Usage: ./cal.rb [-y] [[month] year]
       ./cal.rb [-m month] [year]

また、year, month の値として有効な値を下記に列挙します。
・year
1から9999の整数を表す文字列
・month
1から12の整数を表す文字列
1から12の整数を表す文字列の末尾に'f'または'p'を加えたもの
先頭の3文字が、各月を表す英単語の先頭の3文字に大文字小文字の区別なく一致する任意の文字列
のいずれか

上記はcalコマンドで有効な値と同等になるように定義しています。

yysd94 committed:
ファイル名で直接プログラムを実行できるようにshebangの記述を追加しました。

yysd94 committed:
cal.rbにuserでの実行権限を付加しました。

yysd94 committed:
無効な年の値を入力したときのエラーメッセージの表示崩れを修正しました。

yysd94 committed:
今日の日付がカレンダーに表示される場合に文字色と背景色を入れ替えて表示する機能を追加しました。

yysd94 committed:
ソースコードの可読性の改善を行いました。

・不要なコメントの削除
・変数名を分かりやすい名前に変更
・煩雑な処理を機能ごとにメソッドにまとめて、mainメソッドから呼び出すように変更

yysd94 committed:
ソースコードの不要なコメント・行を削除しました

yysd94 committed:
月と年の有効な入力値を表す正規表現を修正しました。

yysd94 committed:
月の入力値にfまたはpオプションが指定されたときのロジックを修正しました。

yysd94 committed:
Usageの表示メッセージを修正しました。

yysd94 committed:
入力値のエラーメッセージを修正しました。

yysd94 committed:
月の入力値から番号を取得する部分のソースコードを分かりやすい形に微調整しました。

yysd94 committed:
calコマンドと見た目を一致させるため、カレンダーの出力表示の最後に改行を１行追加しました。

yysd94 committed:
Revert "calコマンドと見た目を一致させるため、カレンダーの出力表示の最後に改行を１行追加しました。"

This reverts commit 418af2b993bb9273186d65bbb7e33f5334a00d0a.

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,146 @@
+
+def main
+  # コマンドラインから入力値を取得
+  input_month, input_year = get_input_month_and_year()
```
問題は無いですがrubyの慣習として `get_hoge` / `set_hoge` を付けないので慣れていきましょう。 (替わりに `hoge` / `hoge=` となります)

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,146 @@
+    # 各フォーマットにマッチした場合に応じて、月の番号を取得する
+    if m_m_s
+      table_of_month_name_to_index = {
+        "JAN" => 1, "FEB" => 2, "MAR" => 3, "APR" => 4, "MAY" => 5, "JUN" => 6,
```
こういうhashの定数定義は

```
"JAN" => 1,
"FEB"  => 2,
...
``` 

のように一行一要素にした方が見やすいし編集時もラクですね。

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,146 @@
+  month = nil
+  year = nil
+  if input_year
+    m_y = /^[0-9]{0,}$/.match(input_year)
```
この 一文字を繋げる変数名は分かりづらいので略さない方が良いですね。
ruby界隈での頻出の略し方が分かってくるまではいったん長くても略さない方が良いと思います。

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,146 @@
+  return input_month, input_year
+end
+
+def get_valid_month_and_year(input_month, input_year)
```
このメソッドは長い上、 validationと値の数値への変換が混ざっていて見づらいですね。
今の長さだとそこは分けた方が良いと思います。

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,146 @@
+      else
+        year = input_year.to_i
+      end
+    else
```
こういう後のelseで exitするとネストが必要になりますが以下のように先に評価してしまえばネストが一段減るので見やすくなります。
rubyは以下のような後置 if/unless が使えてシンプルに書けるので使っていきましょう　💪 

```
puts "cal.rb: not a valid year #{input_year}" unless m_y
...
```

yysd94 replied:
後置 if/unless を使うととてもスッキリ書けますね。勉強になりました。ソースコードの他の場所でもうまく使えそうなところがあったので、適用してみます。

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,146 @@
+  indent_length = 3 * first_date_of_month.wday # wdayメソッドの返リ値は0-6 (日曜日が0)
+
+  # カレンダーのヘッダを出力
+  print("#{first_date_of_month.strftime("%B")} #{year}".center(20) + "\n")
```
最後に改行を入れるなら `puts` を使う方がスッキリしますね。

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,146 @@
+  # カレンダーのヘッダを出力
+  print("#{first_date_of_month.strftime("%B")} #{year}".center(20) + "\n")
+  print("Su Mo Tu We Th Fr Sa\n")
+  print("".rjust(indent_length)) # 初日表示部分までの空白を出力
```
ここのコメントは indent_lengthという変数名から分かるので不要です。
なるべくコメントが無くても分かるソースコードにしたいですね。

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,146 @@
+    if date.month == Date.today.month \
+      && date.year == Date.today.year \
+      && date.day == Date.today.day
+      print("\e[7m") # 文字色と背景色を入れ替えて表示(ANSI excape codeを使用)
```
ここのコメントは良いと思います 👍 
他の方法としてメソッドを作ってメソッド名で表現しても良いですね。

yysd94 committed:
変数名・関数名をより慣習的で分かりやすくなるように修正しました。

yysd94 committed:
print文をputsで事足りるものについてはputsに置き換えて簡潔にしました。

yysd94 committed:
意味が分かりにくい変数名を修正しました。

yysd94 committed:
ハッシュの定義の記述箇所をソースコード上で扱いやすくなるように成形しました。

yysd94 committed:
if/unless文の分岐処理の記述を、見やすく簡潔になるように変更しました。

yysd94 committed:
入力値のバリデーションと整数値への変換処理について、可読性が悪い実装となっている問題を修正しました。

入力値のバリデーション処理と整数値への変換処理を１つのメソッド(valid_month_and_year)で同時に行っており、
長くて可読性の悪いロジックとなっている問題がありました。
これを解決するために、役割をより細分化して小さなメソッドに分割し、ロジックの流れが見やすいように改良しました。

・バリデーションと整数値への変換処理の専用メソッドを追加
valid_month? : 月のバリデーションチェックを行うメソッド
valid_year? : 年のバリデーションチェックを行うメソッド
month_name_to_i : 月の名前から整数の番号を取得するメソッド
month_and_year_to_i : 入力値から整数値への変換するメソッド

・月は許容する入力値のフォーマットが複数存在するため、フォーマットごとに専用のバリデーションメソッドを定義
month_name?
month_num?
month_f_option?
month_p_option?

・mainメソッドのロジックを修正

yysd94 commented:
指摘いただいた事項について、すべて修正しました。再度レビューよろしくお願いいたします！

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,158 @@
+
+def month_name?(input_month)
+  month_name = /^(jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)/i.match(input_month)
+  !!month_name
```
これで booleanにできるわけですが、 rubyは false/nil以外は全部 true判定になるのであまり使わないです。
なので matchしないときは nilなので以下で良いですね。

```
def month_name?(input_month)
  /^(jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)/i.match(input_month)
end
```

まあ今のでも良いですが、一行になるのでメソッドにした方が良いかは微妙なところです。

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,158 @@
+  end
+end
+
+def month_name?(input_month)
```
これは `valid_month_name?` とvalidを付けないと意図が分からないです。

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,158 @@
+
+def month_name_to_i(input_month)
+  matched_month_name = /^(jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)/i.match(input_month)
+  table_of_month_name_to_index = {
```
前回指摘漏れました 🙏 
この内容だと定数の方が良いですね。定数にして使うメソッドの直前に移動しましょう。

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,158 @@
+    month, year = month_and_year_to_i(input_month, input_year)
+    # カレンダーを表示
+    display_calender_of_month(month, year) if month && year
+    puts("一年分のカレンダーを表示する機能は未実装です。") if !month && year
```
これも validationする部分でやってしまいたいですね。

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,158 @@
+    # 取得した入力値を、表示すべきmonthとyearの整数値に変換
+    month, year = month_and_year_to_i(input_month, input_year)
+    # カレンダーを表示
+    display_calender_of_month(month, year) if month && year
```
この3つの呼び分けはmonth_and_year_to_i で、 nilだったときにデフォルト値を入れれば必要無いですね。
if文の分岐が増えると読みづらくなっていくので減らすように心がけましょう 💪

u1tnk commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,158 @@
+
+def valid_year?(input_year)
+  return true unless input_year
+  valid_year = /^[0-9]{0,}$/.match(input_year)
```
```
 valid_year = /^[0-9]{0,}$/.match(input_year)
 unless valid_year
 ...
```

この部分ですが、 valid_yearを他で使わないなら以下で大丈夫です。

```
 unless /^[0-9]{0,}$/.match(input_year)
 ...
```
